### ISSUE TITLE
Cleanup and improvement of SPACE and Erosion Deposition

### ISSUE THREAD

**Issue by kbarnhart** [landlab/landlab#708]
in discussions with @cmshobe and @gregtucker about SPACE and Erosion/Deposition we identified some things:

1. Space and ErosionDeposition should share a common base class. This will make it easier if others care to make similar components that define Es and Er differently.

2. Both components should clean up the inits and the calc_ero_rate function.

3. All different methods of instantiation (and the components overall) should be tested.

4. There are some behaviors that we want that neither component does. Two examples:

a) We should be able to provide an "external" input sediment flux. This would be added to `qs_in`. For example a fluvial input of sediment at a grid boundary (alluvial fan???). 

b) When a node is flooded, how much sediment should be deposited? All of it? Or the amount associated with the deposition rate x dt. 

I'm going to submit a PR very shortly that deals with some of these issues.

**Comment by cmshobe** [comment]
I agree with these changes, with the most important in my mind being the cleaning up of the inits.

On #4b:
In lines 603 and 608 don't we already have a rule for flooded or back-tilted nodes in which sediment is added to the alluvial thickness at a rate of $D_s dt$?

**Comment by nreitman** [comment]
@kbarnhart as we discussed I'm really excited about this: "a) We should be able to provide an "external" input sediment flux. This would be added to qs_in. For example a fluvial input of sediment at a grid boundary (alluvial fan???)." and am willing to help with testing!

**Comment by kbarnhart** [comment]
Thanks @nreitman, I'll work on this and tag you when its ready for tests to be made!

**Comment by Basalte** [comment]
I have seen this discussion about ""external" input sediment flux. I'd like to model evolution of alluvial fan with landlab. I am very new in Landlab and I wonder if thsi option  is now avaliable. If not, how model alluvial fan with landlab ?

**Comment by kbarnhart** [comment]
@Basalte thanks for the question. This functionality is not yet present in either SPACE or ErosionDeposition. Like all of landlab, we would welcome pull requests adding functionality. 

The best way to learn about landlab is to check out the documentation (https://landlab.readthedocs.io/) and the library of tutorials (https://landlab.readthedocs.io/en/master/user_guide/tutorials.html). These can be run locally or through binder.

**Comment by gregtucker** [comment]
In principle, it might be possible to do this by creating a modified version of the ErosionDeposition component that adds an external sediment flux field, and adds that external flux as a source term at each grid node (of course, the source could be zero at all but one or a few nodes). It would require studying that component's code carefully first. If I remember right, @alangston used a similar approach in some of her modeling work:

[https://agupubs.onlinelibrary.wiley.com/doi/full/10.1002/2014JF003403](https://agupubs.onlinelibrary.wiley.com/doi/full/10.1002/2014JF003403)

[https://doi.org/10.5194/esurf-6-1-2018](https://doi.org/10.5194/esurf-6-1-2018)

**Comment by Basalte** [comment]
thank you @kbarnhart and @gregtucker  for your answers. Landlab is a great modeling approach. I have already used it for teaching. Following your suggestions, I will therefore try to modify  ErosionDeposition component to include external sediment flux.

**Comment by kbarnhart** [comment]
Glad to hear landlab has been useful in your teaching @Basalte. If you have additional questions about modifying components, definitely reach out through an issue. And if you do add this functionality, we would welcome (and assist) bringing it into the main code repository through a pull request. 

Know also that there is some ongoing work to address a known issues in Space and Erosion Deposition (see Issue #1177). We hope to address this in the next month or two, so I'd recommend checking that issue and keeping your fork update with the main repository's master branch.

**Comment by alangston** [comment]
As @gregtucker remembered correctly, I did include optional external sediment flux in the LateralErosion component. In the lateral erosion component, the variable name for external sediment flux is "qsinlet". You might be able to use the lateral erosion component as a template if you decide to modify ErosionDeposition component similarly. Good luck, @Basalte ! Happy to answer questions if you have them.
